File: server/service/file.go
package service

import (
	"context"
	"errors"
	"fmt"
	"path/filepath"
	"strings"
	"time"

	"server/global"
	"server/model"
	"server/service/oss"
)

type FileService struct{}

var File = new(FileService)

// GetFileList 获取文件列表
func (s *FileService) GetFileList(page, pageSize int, name, ext string) ([]model.SysFile, int64, error) {
	var files []model.SysFile
	var total int64

	db := global.DB.Model(&model.SysFile{}).Where("status = ?", 1)
	if name != "" {
		db = db.Where("name LIKE ?", "%"+name+"%")
	}
	if ext != "" {
		exts := strings.Split(ext, ",")
		if len(exts) == 1 {
			db = db.Where("ext = ?", ext)
		} else {
			db = db.Where("ext IN ?", exts)
		}
	}

	db.Count(&total)
	err := db.Order("id desc").Offset((page-1)*pageSize).Limit(pageSize).Find(&files).Error
	return files, total, err
}

// GetFileByID 根据ID获取文件
func (s *FileService) GetFileByID(id uint) (*model.SysFile, error) {
	var file model.SysFile
	if err := global.DB.First(&file, id).Error; err != nil {
		return nil, err
	}
	return &file, nil
}

// GetFileByMD5 根据MD5获取文件（用于秒传）
func (s *FileService) GetFileByMD5(md5 string) (*model.SysFile, error) {
	var file model.SysFile
	if err := global.DB.Where("md5 = ? AND status = ?", md5, 1).First(&file).Error; err != nil {
		return nil, err
	}
	return &file, nil
}

// CreateFile 创建文件记录
func (s *FileService) CreateFile(file *model.SysFile) error {
	return global.DB.Create(file).Error
}

func (s *FileService) resolveFileStorage(file model.SysFile) (*model.StorageProfile, error) {
	if strings.TrimSpace(file.StorageType) != "" {
		return Storage.GetStorageByType(model.StorageType(file.StorageType))
	}
	return Storage.GetDefaultStorage()
}

func (s *FileService) createFileRecord(filename, key, url, md5 string, fileSize int64, uploaderID uint, storage *model.StorageProfile) *model.SysFile {
	ext := strings.TrimPrefix(filepath.Ext(filename), ".")
	return &model.SysFile{
		Name:          filename,
		Path:          key,
		URL:           url,
		Size:          fileSize,
		Ext:           ext,
		MimeType:      getMimeType(ext),
		MD5:           md5,
		StorageType:   string(storage.Type),
		UploaderID:    uploaderID,
		Status:        1,
	}
}

// DeleteFile 删除文件
func (s *FileService) DeleteFile(id uint) error {
	var file model.SysFile
	if err := global.DB.First(&file, id).Error; err != nil {
		return err
	}

	var userCount int64
	global.DB.Model(&model.SysUser{}).Where("avatar_file_id = ?", id).Count(&userCount)
	if userCount > 0 {
		return errors.New("文件正在被使用，无法删除")
	}

	storage, err := s.resolveFileStorage(file)
	if err != nil {
		return err
	}

	client, err := oss.GetClient(storage)
	if err == nil {
		_ = client.Delete(context.Background(), file.Path)
	}

	return global.DB.Model(&model.SysFile{}).Where("id = ?", id).Update("status", 0).Error
}

// BatchDeleteFiles 批量删除文件
func (s *FileService) BatchDeleteFiles(ids []uint) (int, []string) {
	var successCount int
	var failedMsgs []string

	for _, id := range ids {
		if err := s.DeleteFile(id); err != nil {
			failedMsgs = append(failedMsgs, fmt.Sprintf("ID %d: %s", id, err.Error()))
		} else {
			successCount++
		}
	}

	return successCount, failedMsgs
}

// GenerateFilePath 生成文件存储路径
func (s *FileService) GenerateFilePath(filename string) string {
	ext := filepath.Ext(filename)
	now := time.Now()
	return fmt.Sprintf("%d/%02d/%02d/%d%s", now.Year(), now.Month(), now.Day(), now.UnixNano(), ext)
}

// GetUploadCredential 获取上传凭证
func (s *FileService) GetUploadCredential(filename string) (*oss.UploadCredential, error) {
	storage, err := Storage.GetDefaultStorage()
	if err != nil {
		return nil, fmt.Errorf("获取存储配置失败: %v", err)
	}

	client, err := oss.GetClient(storage)
	if err != nil {
		return nil, fmt.Errorf("创建存储客户端失败: %v", err)
	}

	key := s.GenerateFilePath(filename)
	credential, err := client.GetUploadCredential(context.Background(), key, 30*time.Minute)
	if err != nil {
		return nil, fmt.Errorf("获取上传凭证失败: %v", err)
	}

	return credential, nil
}

// CheckFileMD5 检查文件MD5是否存在（秒传）
func (s *FileService) CheckFileMD5(md5 string) (*model.SysFile, bool) {
	file, err := s.GetFileByMD5(md5)
	if err != nil {
		return nil, false
	}
	return file, true
}

// InitMultipartUpload 初始化分片上传
func (s *FileService) InitMultipartUpload(filename, md5 string, fileSize int64) (*oss.MultipartUpload, *model.StorageProfile, error) {
	storage, err := Storage.GetDefaultStorage()
	if err != nil {
		return nil, nil, fmt.Errorf("获取存储配置失败: %v", err)
	}

	client, err := oss.GetClient(storage)
	if err != nil {
		return nil, nil, fmt.Errorf("创建存储客户端失败: %v", err)
	}

	key := s.GenerateFilePath(filename)
	upload, err := client.InitMultipartUpload(context.Background(), key)
	if err != nil {
		return nil, nil, fmt.Errorf("初始化分片上传失败: %v", err)
	}

	return upload, storage, nil
}

// GetMultipartUploadURLs 获取分片上传URL列表
func (s *FileService) GetMultipartUploadURLs(uploadID, key string, totalParts int, storage *model.StorageProfile) ([]string, error) {
	client, err := oss.GetClient(storage)
	if err != nil {
		return nil, fmt.Errorf("创建存储客户端失败: %v", err)
	}

	urls := make([]string, totalParts)
	for i := 1; i <= totalParts; i++ {
		url, err := client.GetMultipartUploadURL(context.Background(), uploadID, key, i, 30*time.Minute)
		if err != nil {
			return nil, fmt.Errorf("获取分片上传URL失败: %v", err)
		}
		urls[i-1] = url
	}

	return urls, nil
}

// CompleteMultipartUpload 完成分片上传
func (s *FileService) CompleteMultipartUpload(uploadID, key, filename, md5 string, fileSize int64, uploaderID uint, parts []oss.Part) (*model.SysFile, error) {
	storage, err := Storage.GetDefaultStorage()
	if err != nil {
		return nil, fmt.Errorf("获取存储配置失败: %v", err)
	}

	client, err := oss.GetClient(storage)
	if err != nil {
		return nil, fmt.Errorf("创建存储客户端失败: %v", err)
	}

	if err := client.CompleteMultipartUpload(context.Background(), key, uploadID, parts); err != nil {
		return nil, fmt.Errorf("完成分片上传失败: %v", err)
	}

	file := s.createFileRecord(filename, key, client.GetURL(key), md5, fileSize, uploaderID, storage)
	if err := s.CreateFile(file); err != nil {
		return nil, fmt.Errorf("保存文件记录失败: %v", err)
	}

	return file, nil
}

// GetUploadedParts 获取已上传的分片列表
func (s *FileService) GetUploadedParts(uploadID, key string) ([]oss.Part, error) {
	storage, err := Storage.GetDefaultStorage()
	if err != nil {
		return nil, fmt.Errorf("获取存储配置失败: %v", err)
	}

	client, err := oss.GetClient(storage)
	if err != nil {
		return nil, fmt.Errorf("创建存储客户端失败: %v", err)
	}

	return client.ListParts(context.Background(), key, uploadID)
}

// AbortMultipartUpload 取消分片上传
func (s *FileService) AbortMultipartUpload(uploadID, key string) error {
	storage, err := Storage.GetDefaultStorage()
	if err != nil {
		return fmt.Errorf("获取存储配置失败: %v", err)
	}

	client, err := oss.GetClient(storage)
	if err != nil {
		return fmt.Errorf("创建存储客户端失败: %v", err)
	}

	return client.AbortMultipartUpload(context.Background(), key, uploadID)
}

// SaveUploadedFile 保存已上传的文件记录
func (s *FileService) SaveUploadedFile(filename, key, url, md5 string, fileSize int64, uploaderID uint) (*model.SysFile, error) {
	storage, err := Storage.GetDefaultStorage()
	if err != nil {
		return nil, fmt.Errorf("获取存储配置失败: %v", err)
	}

	file := s.createFileRecord(filename, key, url, md5, fileSize, uploaderID, storage)
	if err := s.CreateFile(file); err != nil {
		return nil, fmt.Errorf("保存文件记录失败: %v", err)
	}

	return file, nil
}

// getMimeType 根据扩展名获取MIME类型
func getMimeType(ext string) string {
	mimeTypes := map[string]string{
		"jpg":  "image/jpeg",
		"jpeg": "image/jpeg",
		"png":  "image/png",
		"gif":  "image/gif",
		"bmp":  "image/bmp",
		"webp": "image/webp",
		"svg":  "image/svg+xml",
		"ico":  "image/x-icon",
		"pdf":  "application/pdf",
		"doc":  "application/msword",
		"docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
		"xls":  "application/vnd.ms-excel",
		"xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
		"ppt":  "application/vnd.ms-powerpoint",
		"pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
		"zip":  "application/zip",
		"rar":  "application/x-rar-compressed",
		"7z":   "application/x-7z-compressed",
		"tar":  "application/x-tar",
		"gz":   "application/gzip",
		"mp3":  "audio/mpeg",
		"wav":  "audio/wav",
		"mp4":  "video/mp4",
		"avi":  "video/x-msvideo",
		"mov":  "video/quicktime",
		"wmv":  "video/x-ms-wmv",
		"flv":  "video/x-flv",
		"txt":  "text/plain",
		"html": "text/html",
		"css":  "text/css",
		"js":   "application/javascript",
		"json": "application/json",
		"xml":  "application/xml",
	}

	if mime, ok := mimeTypes[strings.ToLower(ext)]; ok {
		return mime
	}
	return "application/octet-stream"
}

File: web/src/types/file.ts
// 文件信息
export interface FileInfo {
  id: number
  name: string
  path: string
  url: string
  size: number
  ext: string
  mime_type: string
  md5: string
  storage_type: string
  uploader_id: number
  status: number
  reference_count: number
  created_at: string
  updated_at: string
}

// 上传凭证
export interface UploadCredential {
  provider: string
  upload_url: string
  key: string
  expires: number
  headers: Record<string, string>
  form_data: Record<string, string>
  method: string
  preview_url: string
  access_key_id?: string
  access_key_secret?: string
  security_token?: string
  bucket?: string
  region?: string
  endpoint?: string
}

// 分片上传信息
export interface MultipartUpload {
  upload_id: string
  key: string
  bucket?: string
  chunk_size: number
}

// 分片信息
export interface Part {
  part_number: number
  etag: string
  size?: number
}

// 初始化分片上传响应
export interface InitMultipartUploadResponse {
  instant_upload: boolean
  file?: FileInfo
  upload_id?: string
  key?: string
  chunk_size?: number
  total_parts?: number
  upload_urls?: string[]
}

// 上传进度
export interface UploadProgress {
  file: File
  status: 'pending' | 'calculating' | 'uploading' | 'success' | 'error'
  progress: number
  md5?: string
  error?: string
  result?: FileInfo
}

export interface FileMigrationItem {
  file_id: number
  file_name: string
  source_storage_type: string
  target_storage_type: string
  old_url: string
  new_url: string
  action: 'PENDING' | 'SKIP' | 'CONFLICT' | 'MISSING_SOURCE' | 'MIGRATED' | 'WARNING' | 'FAILED'
  message: string
}

export interface FileMigrationResult {
  target_storage_type: string
  total_count: number
  total_size: number
  pending_count: number
  pending_size: number
  skipped_count: number
  skipped_size: number
  conflict_count: number
  conflict_size: number
  missing_source_count: number
  missing_source_size: number
  migrated_count: number
  failed_count: number
  warning_count: number
  items: FileMigrationItem[]
}

export type FileMigrationTaskState = 'SCANNING' | 'RUNNING' | 'SUCCESS' | 'FAILED'

export interface FileMigrationTaskStatus {
  task_id: string
  status: FileMigrationTaskState
  message: string
  scope: FileMigrationScope
  source_storage_type: string
  target_storage_type: string
  total_count: number
  total_size: number
  pending_count: number
  pending_size: number
  skipped_count: number
  skipped_size: number
  conflict_count: number
  conflict_size: number
  missing_source_count: number
  missing_source_size: number
  processed_count: number
  processed_size: number
  migrated_count: number
  failed_count: number
  warning_count: number
  current_file_id: number
  current_file_name: string
  started_at: string
  finished_at: string
  items: FileMigrationItem[]
}

export type FileMigrationScope = 'all' | 'filter' | 'selected'

export interface FileMigrationFilters {
  name?: string
  ext?: string
  referenced?: boolean
}

export interface FileMigrationRequest {
  scope: FileMigrationScope
  ids?: number[]
  source_storage_type: string
  target_storage_type: string
  source_config?: string
  filters?: FileMigrationFilters
}

// 允许的文件类型
export const imageTypes = ['image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/bmp', 'image/svg+xml']
export const documentTypes = [
  'application/pdf',
  'application/msword',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'application/vnd.ms-excel',
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  'application/vnd.ms-powerpoint',
  'application/vnd.openxmlformats-officedocument.presentationml.presentation',
]
export const videoTypes = ['video/mp4', 'video/avi', 'video/quicktime', 'video/x-ms-wmv', 'video/x-flv']
export const audioTypes = ['audio/mpeg', 'audio/wav', 'audio/ogg']
export const archiveTypes = ['application/zip', 'application/x-rar-compressed', 'application/x-7z-compressed', 'application/x-tar', 'application/gzip']

// 文件扩展名映射
export const extMimeMap: Record<string, string> = {
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  png: 'image/png',
  gif: 'image/gif',
  webp: 'image/webp',
  bmp: 'image/bmp',
  svg: 'image/svg+xml',
  pdf: 'application/pdf',
  doc: 'application/msword',
  docx: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  xls: 'application/vnd.ms-excel',
  xlsx: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  ppt: 'application/vnd.ms-powerpoint',
  pptx: 'application/vnd.openxmlformats-officedocument.presentationml.presentation',
  mp4: 'video/mp4',
  mp3: 'audio/mpeg',
  zip: 'application/zip',
  rar: 'application/x-rar-compressed',
  txt: 'text/plain',
  json: 'application/json',
}
